hypothesis-python/src/hypothesis/extra/redis.py

```diff
@@ -36,43 +39,94 @@ def __init__(
         *,
         expire_after: timedelta = timedelta(days=8),
         key_prefix: bytes = b"hypothesis-example:",
+        listener_channel: str = "hypothesis-changes",
     ):
+        super().__init__()
         check_type(Redis, redis, "redis")
         check_type(timedelta, expire_after, "expire_after")
         check_type(bytes, key_prefix, "key_prefix")
+        check_type(str, listener_channel, "listener_channel")
         self.redis = redis
         self._expire_after = expire_after
         self._prefix = key_prefix
+        self.listener_channel = listener_channel
+        self._pubsub: Any = None
 
     def __repr__(self) -> str:
         return (
             f"RedisExampleDatabase({self.redis!r}, expire_after={self._expire_after!r})"
         )
 
     @contextmanager
-    def _pipeline(self, *reset_expire_keys, transaction=False, auto_execute=True):
+    def _pipeline(
+        self,
+        *reset_expire_keys,
+        execute_and_publish=True,
+        event_type=None,
+        to_publish=None,
+    ):
         # Context manager to batch updates and expiry reset, reducing TCP roundtrips
-        pipe = self.redis.pipeline(transaction=transaction)
+        pipe = self.redis.pipeline()
         yield pipe
         for key in reset_expire_keys:
             pipe.expire(self._prefix + key, self._expire_after)
-        if auto_execute:
-            pipe.execute()
+        if execute_and_publish:
+            # pipe.execute returns a value for each operation, which includes
+            # whatever we did in the yield as a prefix, and the n operations from
+            # pipe.expire as a suffix. remove that suffix to get just the prefix.
+            values = pipe.execute()
+            values = values[: -len(reset_expire_keys)]
+            if any(value > 0 for value in values):
```

Comment:
I'm not really clear on the type of `values`; it looks like a sequence of bytestrings but we're comparing to `0`?

Reply:
redis returns the number of rows affected by an operation in `.execute`; I'll clarify in a comment.